### ISSUE TITLE
[question] why bionic needs to save/restore gp for riscv

### ISSUE THREAD

**Issue by unicornx** [google/android-riscv64#84]
hi,
When I read the shadow-call-stack related code, I see in `libc/arch-riscv64/bionic/setjmp.S`, gp will be saved in setjmp while restored in longjmp. But I notice in `libc/bionic/pthread_internal.h`, the comments say “longjmp()/setjmp() between different threads is undefined behavior (and unsupported on Android)”.
So my quesiton is: if we only support longjmp()/setjmp() in one thread, I think gp is always the same for the same thread, why do we need to save/restore the gp?
In short, I have not figured out in what case do we need to save/restore gp for setjmp/longjmp. I think we have the same issue for aarch64. 
Looking forward to your direction and thanks in adv.

**Comment by enh-google** [comment]
> So my quesiton is: if we only support longjmp()/setjmp() in one thread, I think gp is always the same for the same thread, why do we need to save/restore the gp?

think of it like a second stack pointer. if you'd save/restore the regular stack pointer with the locals on it, you'd need to save/restore the shadow stack pointer with the return addresses on it.
